### ISSUE TITLE
Use environment variables for configuration

### ISSUE THREAD

**Issue by phozzy** [kanidm/kanidm#290]
Currently in order to run a server you have to do that using the following command:
`/sbin/kanidmd server -c /data/server.toml`
or something similar.
That means that you have to provide a server configuration using a configuration file. That is nice for static environments, but for more dynamic environments such as k8s or similar ones it is nice to provide configuration using environment variables. It is also has a sense if you use systemd service, since systemd also allows you to inject environment variable in a systemd service.
So we can use one to one mapping of server configuration file parameters to environment variables (almost):
```
export BINDADDRESS="127.0.0.1:8443"
export LDAPBINDADDRESS="127.0.0.1:3636"
export DB_PATH="/tmp/kanidm.db"
export LOG_LEVEL="perffull"
```
And here comes a difference for environment variables for keys and certificates. For example in k8s environment you can keep this sensitive information using k8s-secrets. And use it after in one of the two ways:
* using config-maps, in this scenario those secrets appear inside a container a standard files, so nothing particularly changes here, and you create environment variables containing file paths:
```
export TLS_CA_PATH="/cert/insecure/ca.pem"
export TLS_CERT_PATH="/cert/insecure/cert.pem"
export TLS_KEY_PATH="/cert/insecure/key.pem"
```
the downside of this method is that in case you change the content of your k8s-secret, your deployment may be not aware (this may be not actual now, but it was when I worked with it) of those changes and will not restart your containers in order to reread the content of your k8s-secrets. And here comes the second way.
* using environment variables:
```
export TLS_CA="-----BEGIN CERTIFICATE-----\nblablabal....\nASEASDFA...aEFAE\n...\n...\n-----END CERTIFICATE-----"
export TLS_CERT="-----BEGIN CERTIFICATE-----\nblablabal....\nASEASDFA...aEFAE\n...\n...\n-----END CERTIFICATE-----"
export TLS_KEY_PATH="-----BEGIN PRIVATE KEY-----\n...\n...\n...\n-----END PRIVATE KEY-----"
```
That basically it, of course those variables naming and number can vary, but the idea should be clear.
Is a dependancy for #282

**Comment by phozzy** [comment]
[Here](https://kubernetes.io/docs/concepts/configuration/secret/) is a general information about k8s secrets.

**Comment by ryankurte** [comment]
mm it's a real pain to have to re-template things when deploying services like this eh. i haven't seen a _particularly_ good solution to merging both command line arguments and config files (in rust or, any other language tbqh) :-/

- putting em in `structopt` supports defaults and environmental overrides in a reasonable way. if there's a clean delineation between what would be a command-line / env arg vs. what goes in a config file splitting to cli + config file could work? 
- otherwise you'd need some combination of parsing the cli args to find the config, then the config, then the cli again to override config values, which seems useful but complex and perhaps difficult to rationalise about
- another approach i have taken in the past is to support environmental variable interpolation in the config files but, it's also a bit gnarley and would probably require extension of serde or some interesting newtypes
- defining everything via structopt is viable but, it doesn't scale so well when you have lots of tunable parameters :-/

**Comment by pando85** [comment]
In my opinion, going to environment variable value/path and/or configuration files could be great but it's not a priority.

If we add this must be and aggregation to merge with actual config file, workouts adding complexity and also supporting PATH environment variables because for secrets it is safer to read from files than use environment variables.

**Comment by Firstyear** [comment]
I think there is no rush on this topic, k8s support is a while away, there are a lot of things to do first and the bulk of configuration is stored in the DB anyway so that it's replicated. So we have the luxury of time to think of the right answer here.

**Comment by ryankurte** [comment]
https://github.com/mehcode/config-rs seems like it might work for this

**Comment by Firstyear** [comment]
@yaleman Is this resolved now? I think you added some magic for args to have env versions?

**Comment by yaleman** [comment]
I wouldn't call this done - so far it's only taking env vars for the current command line things (config file path, etc), not the things _in_ the config files... based on the previous PR, we/I could add cli/env flags for those things?

**Comment by giodamelio** [comment]
This would definitely be nice to have! I just want to put forward a format that I just recently came across recently while messing with Authentik and really lik. Basically it allows putting URIs in the static config to make the program read from another source. The two that they support are `env://` and `file://`, but you can imagine adding others easily.

So the toml config might be something like:

```
tls_key = "env://KANIDM_TLS_KEY"
tls_chain = "file:///run/credentials/kanidam_tls_chain"
```

See [Authentik Config](https://goauthentik.io/docs/installation/configuration)

**Comment by Firstyear** [comment]
I think this could add a bit of complexity to the config parser. We wouldn't want people to have to change their configs to match this, and plenty of config options *aren't* from files either. So I think we might be better off with an alternative on the CLI so that users who want env config can flag that on the startup invocation.

**Comment by yaleman** [comment]
Yeah, having lived through log4shell etc, I'd rather use clap with hidden env options and handle that, than have URIs in config files... 😓

**Comment by Firstyear** [comment]
Oh true. That's really true.

**Comment by yaleman** [comment]
^ PR'd some changes for this one 😄

I have an idea for the replication node config bit too, just need to spend some time pondering how to do it cleanly.
